sourceinfo/source_code_info.go

```diff
@@ -29,37 +29,109 @@ import (
 
 	"github.com/bufbuild/protocompile/ast"
 	"github.com/bufbuild/protocompile/internal"
-	"github.com/bufbuild/protocompile/options"
 )
 
+// OptionIndex is a mapping of AST nodes that define options to corresponding
+// paths into the containing file descriptor. The path is a sequence of field
+// tags and indexes that define a traversal path from the root (the file
+// descriptor) to the resolved option field. The info also includes similar
+// information about child elements, for options whose values are composite
+// (like a list or message literal).
+type OptionIndex map[*ast.OptionNode]*OptionSourceInfo
```

Comment:
This got moved out of the `options` package (previously `options.Index`) into here to avoid an import cycle if we want to be able to pass this type into `linker.ValidateOptions`. (Since the `options` package imports `linker`, we couldn't have a type in a `linker` method signature refer back to the `options` package.)